airflow/models/taskinstance.py

```diff
@@ -1760,16 +1758,16 @@ def _handle_reschedule(
     ti.end_date = timezone.utcnow()
     ti.set_duration()
 
-    # Lock DAG run to be sure not to get into a deadlock situation when trying to insert
-    # TaskReschedule which apparently also creates lock on corresponding DagRun entity
-    with_row_locks(
-        session.query(DagRun).filter_by(
-            dag_id=ti.dag_id,
-            run_id=ti.run_id,
-        ),
-        session=session,
-    ).one()
-    # Log reschedule request
+    # set state
+    ti.state = TaskInstanceState.UP_FOR_RESCHEDULE
+
+    ti.clear_next_method_args()
+
+    session.merge(ti)
+    session.commit()
```

Comment:
@potiuk if anyone would have concerns with this, i suspect it would be you.  will also tag @jscheffl for good measure.

Reply:
The only concern I have that it looks like Airflow 2 change only. The interface there will change and rescheduling should be handled in the "API" component, not in the worker, so this code does not really matter and will be heavily changed when task sdk is implemented. Am I right @ashb ?

Reply:
Some more comments. Generally with Airflow 3 the "transaction" boundaries for any operation on the database will have to change heavily - because basically when you reach out to the API component, there is no way to make several calls in the same transaction - so the transaction boundary will be necessarily smaller than every API call implemented (this was one of the major issues with the "internal-api" approach that we tried to maintain the same transaction boundaries "with DB" and "without DB" and it led to some spaghetti code - where we had to move related DB code from one transaction together in a single `@internal_api` call.

Here, I am not sure how "reschedule" will be converted to the new API - and where transaction boundaries will be, but in any case I think this code will be looking quite differently.

Reply:
Here's my thought: we don't need the stronger lock, so better to use the weaker lock.

It does not only affect reschedule, but any insert on a table that has an FK to the locked table.  We use this somewhat broadly.

Reply:
I recalled something. I vaguely recal there was some problem with mysql that FOR UPDATE was actually locking the FKs connected - and it caused deadlocks - because those indexes could be locked in other transactions in different sequence. I think this is the main reason why we have this strong lock here.

But if we make the change for postgres only, then I guess that one is not a problem.

Reply:
No concerns because of that.

Reply:
@potiuk yes, you are correct, this lock was added to address a report of deadlock _(added by you, incidentally :) )_. and i think user was mysql.   but it would have worked the same way on postgres.  postgres FKs work the same way.  when you insert it has to lock the row with the PK that you are referencing so you don't end up with a reference to a row that does not exist.

postgres is unique in that it provides a way to say "i'm not going to delete or change the pk of this row" which means the insert can go through knowing the row won't disappear.  that's the purpose of this lock mode.

but anyway, so yes in a sense this PR is "postgres only" in that we only change the lock used for concurrency management.  but since mysql doesn't have that mode, i had to still ensure we protect against deadlock in mysql.  and the way i did that was to introduce a commit immediately before inserting taskreschedule.  this will avoid deadlock in mysql because there will not be any other operation in the transaction where the TR record is inserted.  and since there will be nothing else in the tran, then a deadlock will be impossible.  i do not see any downside to committing before.

Reply:
The only potential drawback is that if the system crashes right after commit, the task status will be updated, but TaskReschedule will not be created. Likelihood of it is small, but I think it's worth to at least do a mental exercise of what happens then and whether it's worth it.

Now, when I think of it, it's little worrying - because it actually happens in "_run_raw_task" - which means that it has **some** probabilty to happen (OOM and others). 

BTW. This part should look pretty differently for Airflow 3 with AIP-72 because reschedule will not happen in task but in the API component most likely, so I am not quite sure if that is targetted to be only back-ported? 

@ashb - any concerns and comments here?

Reply:
Yeah I thought about that.  To me it seems a good tradeoff.  Let's discuss.

First, the probability would seem to be quite low. Because we're talking about a very small time window. 

```
1. session.commit
2. session.add
3. session.commit
```
the time is between 1 and 3. 

if 3 is blocked due to locking that could be less insignifigant.

but i think either way it seems we arrive at a better outcome.

why?

because let's imagine the worker blows up between 1 and 3.   What is the consequence?

I believe the consequence would be that no TR record is inserted although the state is set properly.  

That scenario looks to be handled here in the relevant dep: https://github.com/apache/airflow/blob/85fd0e1102d664337d6bb08e590979867372f61d/airflow/ti_deps/deps/ready_to_reschedule.py#L85

So the sensor would be resumed probably more quickly than it thought it would.

But with the before code, it's actually worse what happens.  With the old code if we imagine worker blows up at that same time, then the task does not report state correctly, it would just be an OOM task death which would be a zombie that would be detected by scheduler and the sensor would not resume poking until after retries if configured.

So commiting there that seems better to me anyway, if you assume that it blows up immediately after.

And, commiting there means no deadlock concern, and it means we don't need to hold the table with a select for update, which holds the table for significantly longer, and if you have lots of other sensors in same dag, it can cause trouble.

Reply:
Quite agree. No concerns any more.